packages/nextjs/src/utils/withSentry.ts

```diff
@@ -74,7 +74,57 @@ export const withSentry = (origHandler: NextApiHandler): WrappedNextApiHandler =
+
+        // Temporarily mark the response as finished, as a hack to get nextjs to not complain that we're coming back
+        // from the handler successfully without `res.end()` having completed its work.  This is necessary (and we know
+        // we can do it safely) for a few reasons:
+        //
+        // - Normally, `res.end()` is sync and completes before the request handler returns, as part of the handler
+        //   sending data back to the client. As soon as the handler is done, nextjs checks to make sure that the
+        //   response is indeed finished. (`res.end()` signals this by setting `res.finished` to `true`.) If it isn't,
+        //   nextjs complains.("Warning: API resolved without sending a response for <url>.")
+        //
+        // - In order to prevent the lambda running the route handler from shutting down before we can send events to
+        //   Sentry, we monkeypatch `res.end()` so that we can call `flush()`, wait for it to finish, and only then
+        //   allow the response to be marked complete. This turns the normally-sync `res.end()` into an async function,
+        //   which isn't awaited because it's assumed to still be sync. As a result, nextjs runs the aforementioned
+        //   check before the now-async `res.end()` has had a chance to set `res.finished = false`, and therefore thinks
+        //   there's a problem when there's not.
+        //
+        // - In order to trick nextjs into not complaining, we can set `res.finished` to `true` before exiting the
+        //   handler. If we do that, though, `res.end()` gets mad because it thinks *it* should be the one to get to
+        //   mark the response complete. We therefore need to flip it back to `false` 1) after nextjs's check but 2)
+        //   before the original `res.end()` is called.
+        //
+        // - The second part is easy - we control when the original `res.end()` is called, so we can do the flipping
+        //   right beforehand and `res.end()` will be none the wiser.
+        //
+        // - The first part isn't as obvious. How do we know we won't end up with a race condition, such that the
+        //   flipping to `false` might happen before the check, negating the entire purpose of this hack? Fortunately,
+        //   before it's done, our async `res.end()` wrapper has to await a `setImmediate()` callback, guaranteeing its
+        //   run lasts at least until the next event loop. The check, on the other hand, happens synchronously,
+        //   immediately after the request handler (so in the same event loop). So as long as we wait to flip
+        //   `res.finished` back to `false` until after the `setImmediate` callback has run, we know we'll be safely in
+        //   the next event loop when we do so.
+        //
+        // And that does it! Nextjs doesn't complain about an unfinished response, `res.end()` doesn’t complain about an
+        // already-finished response, and everyone wins.
```

Comment:
I'd remove this. The explanations above are really good; but also long, and I find this makes them unnecessarily longer.

Reply:
I reworded a little, but didn't take it out because I put it there exactly BECAUSE the comment is long, and I think it's helpful to have a one-sentence summary/reminder-of-what-the-heck-we-were-even-talking-about to wrap it up.